core/src/main/java/org/apache/calcite/rel/type/RelDataTypeFactoryImpl.java

```diff
@@ -546,6 +546,33 @@ public boolean isNullable() {
       return nullable;
     }
 
+    public int getPrecision() {
+      final SqlTypeName typeName =
+          JavaToSqlTypeConversionRules.instance().lookup(clazz);
+      if (typeName != null) {
+        return typeSystem.getDefaultPrecision(typeName);
+      }
+      return PRECISION_NOT_SPECIFIED;
+    }
+
+    public int getScale() {
+      final SqlTypeName typeName =
```

Comment:
Why add these 2 methods ? Where they are used ?

Reply:
They are used in validation phase for **JavaType(BigDecimal)**.
For example:
```
select "primitiveByte" + "bigDecimal" as c
from "s"."everyTypes"
where "primitiveByte" <> 0
```
We will get the below stacktrace:

```
java.lang.AssertionError
	at org.apache.calcite.rel.type.RelDataTypeSystem.deriveDecimalPlusType(RelDataTypeSystem.java:151)
	at org.apache.calcite.sql.type.ReturnTypes.lambda$static$6(ReturnTypes.java:516)
	at org.apache.calcite.sql.type.SqlTypeTransformCascade.inferReturnType(SqlTypeTransformCascade.java:54)
	at org.apache.calcite.sql.type.SqlReturnTypeInferenceChain.inferReturnType(SqlReturnTypeInferenceChain.java:56)
	at org.apache.calcite.sql.SqlOperator.inferReturnType(SqlOperator.java:470)
	at org.apache.calcite.sql.SqlOperator.validateOperands(SqlOperator.java:437)
	at org.apache.calcite.sql.SqlOperator.deriveType(SqlOperator.java:516)
	at org.apache.calcite.sql.SqlBinaryOperator.deriveType(SqlBinaryOperator.java:144)
	at org.apache.calcite.sql.validate.SqlValidatorImpl$DeriveTypeVisitor.visit(SqlValidatorImpl.java:5590)
	at org.apache.calcite.sql.validate.SqlValidatorImpl$DeriveTypeVisitor.visit(SqlValidatorImpl.java:5577)
	at org.apache.calcite.sql.SqlCall.accept(SqlCall.java:139)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.deriveTypeImpl(SqlValidatorImpl.java:1677)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.deriveType(SqlValidatorImpl.java:1662)
	at org.apache.calcite.sql.SqlAsOperator.deriveType(SqlAsOperator.java:133)
	at org.apache.calcite.sql.validate.SqlValidatorImpl$DeriveTypeVisitor.visit(SqlValidatorImpl.java:5590)
	at org.apache.calcite.sql.validate.SqlValidatorImpl$DeriveTypeVisitor.visit(SqlValidatorImpl.java:5577)
	at org.apache.calcite.sql.SqlCall.accept(SqlCall.java:139)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.deriveTypeImpl(SqlValidatorImpl.java:1677)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.deriveType(SqlValidatorImpl.java:1662)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.expandSelectItem(SqlValidatorImpl.java:465)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validateSelectList(SqlValidatorImpl.java:4068)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validateSelect(SqlValidatorImpl.java:3375)
```

Reply:
As the comments in `ReflectiveSchemaTest` in the PR.
https://github.com/apache/calcite/pull/1375/files

Reply:
I still can not see any invocation from the stack trace you pasted. Maybe you can add the comments above the methods.

Reply:
If there are some operands with Decimal types in RexCall, its return type is Decimal, and the validator will calculate the result's precision and scale during inferring return type.

However, as we can see in **everyTypes**, when the column is a reflected JavaType(BigDecimal), its `(precision, scale)=(PRECISION_NOT_SPECIFIED, SCALE_NOT_SPECIFIED) = (-1, Integer.MIN_VALUE)`.

During the calculation for the return type's precision and scale in `RelDataTypeSystem.deriveDecimalPlusType`, it leads to Assertions (e.g., `assert precision > 0`).

Reply:
@danny0405 thanks for your suggestion, I will add some comments for this change.

Reply:
In CALCITE-2302, i get a default decimal type for the Java type and use it's precision/scale, i think you can do this also. [1]

[1] https://github.com/apache/calcite/pull/706/files#diff-eb76e2b906a6e4edbcc85056ad58ce75R143

Reply:
Yes, we can do it in `RelDataTypeSystem`.  But we may also need to calculate JavaType's precision/scale in `deriveDecimalMultiplyType`,` deriveDecimalPlusType` and `deriveDecimalDivideType`.